### PR TITLE
SNOW-715634: Support authentication with OAuth tokens

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-jy**
- [x] Implemented Oauth token authentication
- [x] Added a new field for the auth token called oauth_token in SF_CONNECT

### CODE REVIEW

File: lib/client.c [hunk 1]
```diff
@@ -802,7 +813,7 @@ SF_STATUS STDCALL snowflake_connect(SF_CONNECT *sf) {
         // connection object doesn't exists
         return SF_STATUS_ERROR_CONNECTION_NOT_EXIST;
     }
-    if (sf->token || sf->master_token) {
+    if (!is_token_required(sf) && (sf->token || sf->master_token)) {
```
Comment (sfc-gh-ext-simba-hx) [inline lib/client.c:816]: @sfc-gh-ext-simba-jy don't reuse the existing token in SF_CONNECT, that's the session token used in the header of every single request. You can check ODBC implementation as a reference, there are many different tokens in connection class, (token, session token, master token, id token, mfa token).

File: lib/client.c [hunk 2]
```diff
@@ -1072,6 +1088,9 @@ SF_STATUS STDCALL snowflake_set_attribute(
         case SF_CON_AUTHENTICATOR:
             alloc_buffer_and_copy(&sf->authenticator, value);
             break;
+        case SF_CON_OAUTH_TOKEN:
```
Comment (sfc-gh-ext-simba-hx) [inline lib/client.c:1091]: Please add to snowflake_get_attribute() as well.

File: lib/connection.c [hunk 3]
```diff
@@ -173,6 +173,10 @@ cJSON *STDCALL create_auth_json_body(SF_CONNECT *sf,
     if (sf->password && *(sf->password)) {
         snowflake_cJSON_AddStringToObject(data, "PASSWORD", sf->password);
     }
+
+    if (sf->oauth_token && *(sf->oauth_token)) {
```
Comment (sfc-gh-ext-simba-hx) [inline lib/connection.c:177]: Here we need to check authenticator type. "TOKEN" in login request has different meaning with different authenticators.
https://github.com/snowflakedb/snowflake-odbc/blob/master/Source/Platform/Connection.cpp#L605

Comment (sfc-gh-ext-simba-jy) [reply lib/connection.c]: Done.

File: lib/client.c (hunk 2 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/client.c:1091]: Done.

File: tests/test_unit_oauth.c [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    assert_int_not_equal(status, SF_STATUS_SUCCESS);
+    SF_ERROR_STRUCT* error = snowflake_error(sf);
+    assert_int_equal(error->error_code, SF_STATUS_ERROR_BAD_CONNECTION_PARAMS);
+    snowflake_term(sf);
+}
+
+/*
+ * Test the request body with the oauth connection.
+ */
+void test_json_data_in_Oauth(void** unused) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-jy) [reply tests/test_unit_oauth.c]: Done.

File: tests/test_oauth.c [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    port = getenv("SNOWFLAKE_TEST_PORT");
+    if (port) {
+        snowflake_set_attribute(sf, SF_CON_PORT, port);
+    }
+    protocol = getenv("SNOWFLAKE_TEST_PROTOCOL");
+    if (protocol) {
+        snowflake_set_attribute(sf, SF_CON_PROTOCOL, protocol);
+    }
+    char* token = "<Pass your token here>";
+    snowflake_set_attribute(sf, SF_CON_OAUTH_TOKEN, token);
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-jy) [reply tests/test_oauth.c:31]: I accidentally removed it when I updated the PR. I revert it. Thank you.

File: tests/test_oauth.c [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#include "utils/test_setup.h"
+
+/**
+ * Test connection with OAuth authentication.
+ */
+void test_connect_with_oauth(void **unused) {
+    SF_CONNECT *sf = snowflake_init();
+    snowflake_set_attribute(sf, SF_CON_ACCOUNT,
+                            getenv("SNOWFLAKE_TEST_ACCOUNT"));
+    //snowflake_set_attribute(sf, SF_CON_USER, getenv("SNOWFLAKE_TEST_USER"));
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-jy) [reply tests/test_oauth.c]: Removed.

File: cpp/lib/Authenticator.cpp [hunk 7]
```diff
@@ -123,6 +127,21 @@ extern "C" {
...
+    {
+        cJSON* data = snowflake_cJSON_GetObjectItem(body, "data");
+        if (!data)
+        {
+            data = snowflake_cJSON_CreateObject();
+            snowflake_cJSON_AddItemToObject(body, "data", data);
+        }
+
+        snowflake_cJSON_DeleteItemFromObject(data, "AUTHENTICATOR");
+        snowflake_cJSON_AddStringToObject(data, "AUTHENTICATOR", "OAUTH");
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-jy) [reply cpp/lib/Authenticator.cpp]: Updated.

**Review by sfc-gh-dstempniak — APPROVED** [review]
LGTM

File: lib/client.c [hunk 8]
```diff
@@ -514,6 +525,9 @@ _snowflake_check_connection_parameters(SF_CONNECT *sf) {
         log_debug("jwt_timeout: %d", sf->jwt_timeout);
         log_debug("jwt_cnxn_wait_time: %d", sf->jwt_cnxn_wait_time);
     }
+    if (AUTH_OAUTH == auth_type) {
+        log_debug("oauth_token: %s", sf->oauth_token ? "****" : sf->oauth_token);
```
Comment (sfc-gh-dprzybysz) [inline lib/client.c:529]: in other drivers we want to log only information that sensitive data is provided or not, please change to:
```
log_debug("oauth_token: %s", sf->oauth_token ? "provided" : "not provided");
```

File: lib/error.h [hunk 9]
```diff
@@ -38,6 +38,7 @@ void STDCALL copy_snowflake_error(SF_ERROR_STRUCT *dst, SF_ERROR_STRUCT *src);
 #define ERR_MSG_GONE_SESSION "The session no longer exists on the server. Please reconnect"
 #define ERR_MSG_AUTHENTICATOR_UNSUPPORTED "unsupported authenticator"
 #define ERR_MSG_PRIVKEYFILE_PARAMETER_IS_MISSING "priv_key_file parameter is missing"
+#define ERR_MSG_TOKEN_PARAMETER_IS_MISSING "token parameter is missing"
```
Comment (sfc-gh-dprzybysz) [inline lib/error.h:41]: let's add OAuth at the beginning of the error message

File: tests/test_manual_connect.c [hunk 10]
```diff
@@ -1,19 +1,53 @@
-/*
- * Copyright (c) 2018-2024 Snowflake Computing, Inc. All rights reserved.
- */
+//
```
Comment (sfc-gh-dprzybysz) [inline tests/test_manual_connect.c:1]: why do we need to change the comments format here?

File: lib/client.c (hunk 8 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/client.c]: Updated.

File: lib/error.h (hunk 9 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/error.h:41]: I added 'OAuth' in both the variable and the message.

File: tests/test_manual_connect.c (hunk 10 above)
Comment (sfc-gh-ext-simba-jy) [reply tests/test_manual_connect.c]: I reverted back.

File: tests/test_manual_connect.c [hunk 11]
```diff
@@ -117,14 +151,15 @@ void test_connect_with_duo_passcodeInPassword(void** unused)
     snowflake_term(sf);
 }
 
-int main(void)
-{
+
+int main(void) {
```
Comment (sfc-gh-dprzybysz) [inline tests/test_manual_connect.c:155]: let's apply the same formatting rules that we have in other places in the library

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: tests/test_manual_connect.c (hunk 11 above)
Comment (sfc-gh-ext-simba-jy) [reply tests/test_manual_connect.c]: Done.

File: tests/test_manual_connect.c [hunk 12]
```diff
@@ -1,19 +1,57 @@
...
+    port = getenv("SNOWFLAKE_TEST_PORT");
+    if (port) 
+    {
+        snowflake_set_attribute(sf, SF_CON_PORT, port);
+    }
+    protocol = getenv("SNOWFLAKE_TEST_PROTOCOL");
+    if (protocol) 
+    {
+        snowflake_set_attribute(sf, SF_CON_PROTOCOL, protocol);
+    }
```
Comment (sfc-gh-jszczerbinski) [inline tests/test_manual_connect.c:32]: Please add setting role based on env, it's common that we would like to select non-default role in OAuth tests:
```c
role = getenv("SNOWFLAKE_TEST_ROLE");
if (role)
{
   snowflake_set_attribute(sf, SF_CON_ROLE, role);
}
```

File: tests/test_manual_connect.c [hunk 13]
```diff
@@ -117,14 +155,15 @@ void test_connect_with_duo_passcodeInPassword(void** unused)
     snowflake_term(sf);
 }
 
-int main(void)
+int main(void) 
 {
     initialize_test(SF_BOOLEAN_FALSE);
     const struct CMUnitTest tests[] = {
```
Comment (sfc-gh-jszczerbinski) [inline tests/test_manual_connect.c:161]: Can we use env var (ex. "SNOWFLAKE_MANUAL_TEST_TYPE") to select which test we want to run?

**Review by sfc-gh-jszczerbinski — APPROVED** [review]
Please apply the UX changes to manual tests. LGTM

File: tests/test_manual_connect.c (hunk 13 above)
Comment (sfc-gh-ext-simba-jy) [reply tests/test_manual_connect.c]: Done.

**Pull request merged**